### PR TITLE
metric[engine]: framework support metric interface

### LINKED ISSUE #5496: BaseMaster/BaseWorker support  multi-projects/tasks metric interface

### PR DESCRIPTION
**Pull request by maxshuang**
### What problem does this PR solve?

Issue Number: close  #5496

### What is changed and how it works?

1. Offer MetricFactory() method in BaseMaster/BaseWorker/BaseJobMaster
2. pass worker type when creating master/worker

```release-note
None
```

### CODE REVIEW

File: engine/lib/master.go [hunk 1]
```diff
@@ -188,6 +191,7 @@ func NewBaseMaster(
 	ctx *dcontext.Context,
 	impl MasterImpl,
 	id libModel.MasterID,
+	tp libModel.WorkerType,
```
Comment (amyangfei) [inline engine/lib/master.go:194]: What about passing `WorkerType` in `dcontext.Context`

File: engine/lib/common.go [hunk 2]
```diff
@@ -66,3 +66,37 @@ type WorkerHandle = master.WorkerHandle
...
+// TODO: let user register a unique identifier for the metric prefix
+func WorkerTypeForMetric(t WorkerType) string {
+	switch t {
+	case JobManager:
+		// jobmanager is the framework level job
+		return ""
+	case CvsJobMaster:
+		return "cvs"
+	case CvsTask:
+		return "cvs"
```
[the remark this answers is not in the document]
Comment (maxshuang) [reply engine/lib/common.go]: resolved in latest commit

File: engine/lib/master.go (hunk 1 above)
Comment (maxshuang) [reply engine/lib/master.go:194]: What's the principle to pass some params in `dcontext.Context` ?

Comment (amyangfei) [reply engine/lib/master.go:194]: There are no clear rules, I think passing worker type is reasonable.

**Review by liuzix — APPROVED** [review]

**Pull request merged**